src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/VersionConverter.cs

```diff
@@ -1,25 +1,111 @@
+                {
+                    throw ThrowHelper.GetFormatException(DataType.Version);
+                }
+
+                Span<byte> stackSpan = stackalloc byte[isEscaped ? MaximumEscapedVersionLength : MaximumVersionLength];
+                reader.ValueSequence.CopyTo(stackSpan);
+                source = stackSpan.Slice(0, (int)reader.ValueSequence.Length);
+            }
+            else
+            {
+                source = reader.ValueSpan;
+
+                if (!JsonHelpers.IsInRangeInclusive(source.Length, MinimumVersionLength, maxLength))
+                {
+                    throw ThrowHelper.GetFormatException(DataType.Version);
+                }
+            }
+
+            if (isEscaped)
+            {
+                int backslash = source.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(backslash != -1);
+
+                Span<byte> sourceUnescaped = stackalloc byte[MaximumEscapedVersionLength];
+
+                JsonReaderHelper.Unescape(source, sourceUnescaped, backslash, out int written);
+                Debug.Assert(written > 0);
+
+                source = sourceUnescaped.Slice(0, written);
+                Debug.Assert(!source.IsEmpty);
+            }
+
+            byte firstChar = source[0];
+            byte lastChar = source[source.Length - 1];
+            if (!JsonHelpers.IsDigit(firstChar) || !JsonHelpers.IsDigit(lastChar))
+            {
+                // Disallow leading and trailing whitespaces
+                // since Version.Parse allows them
+                throw ThrowHelper.GetFormatException(DataType.Version);
+            }
+
+            Span<char> charBuffer = stackalloc char[MaximumVersionLength];
+            int writtenChars = JsonReaderHelper.s_utf8Encoding.GetChars(source, charBuffer);
+            if (Version.TryParse(charBuffer.Slice(0, writtenChars), out Version? result))
+            {
+                return result;
+            }
+#else
             string? versionString = reader.GetString();
+            if (!string.IsNullOrEmpty(versionString) && (!char.IsDigit(versionString[0]) || !char.IsDigit(versionString[versionString.Length - 1])))
+            {
+                throw ThrowHelper.GetFormatException(DataType.Version);
+            }
             if (Version.TryParse(versionString, out Version? result))
             {
                 return result;
             }
-
+#endif
             ThrowHelper.ThrowJsonException();
             return null;
         }
 
         public override void Write(Utf8JsonWriter writer, Version value, JsonSerializerOptions options)
         {
+#if BUILDING_INBOX_LIBRARY
+            Span<char> span = stackalloc char[MaximumVersionLength];
+            bool formattedSuccessfully = value.TryFormat(span, out int charsWritten);
+            Debug.Assert(formattedSuccessfully && charsWritten >= MinimumVersionLength);
+            writer.WriteStringValue(span.Slice(0, charsWritten));
+            return;
```

Comment:
nit: remove these return statements, unless they are necessary?

Reply:
They aren't necessary. Removed them in a5a8043